src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,226 @@
+section \<open>Chain-Sync Mini-Protocol\<close>
+
+text \<open>
+  The theory in this section implements the chain synchronization mini-protocol.
+\<close>
+
+theory "Ouroboros-Mini_Protocols-Chain_Sync"
+  imports
+    "Ouroboros-Mini_Protocols"
+    "HOL-Library.BNF_Corec"
+begin
+
+subsection \<open>Cryptographic Primitives\<close>
+
+typedecl hash
+
+axiomatization
+where
+  hash_linear_order: "OFCLASS(hash, linorder_class)"
+
+instance hash :: linorder
+  by (fact hash_linear_order)
+
+subsection \<open>Blocks\<close>
+
+datatype block = Block (hash: \<open>hash\<close>)
+
+subsection \<open>Chains\<close>
+
+type_synonym chain = "block list"
```

Comment:
It might be debatable whether we should introduce an extra name for the type of chains if chains are really just lists of blocks. However, when realizing my above suggestions, we won’t have chains always but perhaps only lists of headers, and this in turn would mean we would have to maintain hashes separately. As a result, what we would really have to maintain would be lists of pairs of hashes and associated data items. Introducing a new name for this new kind of structure certainly makes some sense. That said, we might want to introduce a type for combinations of one hash and one data item and then use a plain list type for lists of such combinations.